### PR TITLE
System Fetcher with Operations

### PR DESCRIPTION
**Pull request by dzahariev**
**Description**

Changes proposed in this pull request:
- System Fetcher uses Operation framework
- Unit tests for System Fetcher service
- Deployment descriptors to be adjusted
- E2E tests

**Pull Request status**
- [x] Implementation
- [x] Unit tests
- [x] Integration tests
- [x] `chart/compass/values.yaml` is updated 
- [x] Mocks are regenerated, using the automated script

### CODE REVIEW

File: components/director/internal/systemfetcher/operation_maintainer.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+		return nil, nil, errors.Wrap(err, "while getting tenants")
+	}
+
+	desiredStateOperations := make([]*model.OperationInput, 0)
+	for _, tenant := range tenants {
+		switch tenant.Type {
+		case tenantpkg.Customer, tenantpkg.Account:
+			op, err := om.tenantToOperation(tenant)
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "while creating operation from account tenant %q", tenant.ID)
```
Comment (dragobt) [inline components/director/internal/systemfetcher/operation_maintainer.go:103]: `while creating operation from account tenant` -> `while creating operation from account or customer tenant`

File: components/director/internal/domain/systemssync/repository.go [hunk 2]
```diff
@@ -40,10 +41,10 @@ func NewRepository(conv EntityConverter) *repository {
 }
 
 // List returns all system sync timestamps from database
```
Comment (dragobt) [inline components/director/internal/domain/systemssync/repository.go:43]: `List` -> `ListByTenant`

File: components/director/internal/domain/tenant/service.go [hunk 3]
```diff
@@ -566,6 +566,19 @@ func (s *service) ensureTenantExists(ctx context.Context, id string) error {
 	return nil
 }
 
+func (s *service) ExistsByExternalTenant(ctx context.Context, externalTenant string) error {
```
Comment (dragobt) [inline components/director/internal/domain/tenant/service.go:569]: add go doc

**Review by dragobt — CHANGES_REQUESTED** [review]

File: components/director/internal/systemfetcher/service.go [hunk 4]
```diff
@@ -141,149 +132,147 @@ type tenantSystems struct {
...
-	tenants, err := s.listTenants(ctx, tenantType)
+	systems, err := s.systemsAPIClient.FetchSystemsForTenant(ctx, tenant, systemSynchronizationTimestamps)
 	if err != nil {
-		return errors.Wrap(err, "failed to list tenants")
+		return errors.Wrap(err, fmt.Sprintf("failed to fetch systems for tenant %s of type %s", tenant.ExternalTenant, tenant.Type))
+	}
+
+	log.C(ctx).Infof("Found %d systems for tenant %s of type %s", len(systems), tenant.ExternalTenant, tenant.Type)
+	if len(s.config.VerifyTenant) > 0 {
+		log.C(ctx).Infof("Systems: %#v", systems)
```
[the remark this answers is not in the document]
Comment (dzahariev) [reply components/director/internal/systemfetcher/service.go:166]: Used in tests and troubleshooting.

File: components/director/cmd/systemfetcher/main.go [hunk 5]
```diff
@@ -478,27 +612,9 @@ func createSystemFetcher(ctx context.Context, cfg config, cfgProvider *configpro
...
-	if err := json.Unmarshal([]byte(cfg.TemplateConfig.PlaceholderToSystemKeyMappings), &placeholdersMapping); err != nil {
-		return nil, errors.Wrapf(err, "while unmarshaling placeholders mapping")
-	}
-
-	if err := calculateTemplateMappings(ctx, cfg, tx, appTemplateSvc, placeholdersMapping); err != nil {
-		return nil, errors.Wrap(err, "failed while calculating application templates mappings")
-	}
-
-	templateRenderer, err := systemfetcher.NewTemplateRenderer(appTemplateSvc, appConverter, cfg.TemplateConfig.OverrideApplicationInput, placeholdersMapping)
+	templateRenderer, err := reloadTemplates(ctx, cfg, tx)
```
[the remark this answers is not in the document]
Comment (dzahariev) [reply components/director/cmd/systemfetcher/main.go:615]: I will keep this one here, as we are not aware of all usages of loaded resources.  Also this information can be needed for System Fetcher proper initialisation.

File: components/director/internal/systemfetcher/handler.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+			log.C(ctx).WithError(err).Errorf("Loading Operation for System Fetcher data aggregation failed")
+			http.Error(writer, "Loading Operation for System Fetcher data aggregation failed", http.StatusInternalServerError)
+			return
+		}
+
+		log.C(ctx).Infof("Operation with TenantID %q does not exist. Trying to create...", payload.TenantID)
+
+		// Check if the provided tenant exists.
+		businessTenantMappingID := payload.TenantID
+		err := h.existsBusinessTenantMappingByID(ctx, payload.TenantID)
```
[the remark this answers is not in the document]
Comment (dzahariev) [reply components/director/internal/systemfetcher/handler.go:79]: It works with internal and also fallbacks to external id if internal is not found. Anyway the operations are always stored with information about internal tenant ID.

File: components/director/internal/systemfetcher/handler.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+		h.onDemandChannel <- opID
+
+		writer.WriteHeader(http.StatusOK)
+		return
+	}
+
+	if err = h.opMgr.RescheduleOperation(ctx, operation.ID); err != nil {
+		log.C(ctx).WithError(err).Errorf("Failed to reschedule operation with ID %s", operation.ID)
+		http.Error(writer, "Scheduling Operation for System Fetcher data aggregation failed", http.StatusInternalServerError)
+		return
```
Comment (dragobt) [inline components/director/internal/systemfetcher/handler.go:142]: There is no unit test if `RescheduleOperation(...)` fails.

File: components/director/internal/systemfetcher/service.go [hunk 8]
```diff
@@ -141,149 +131,147 @@ type tenantSystems struct {
...
-						syncTimestamp: currentTime,
-					}
-				}
-			}(t)
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	syncTimestamps, err := s.systemsSyncService.ListByTenant(ctx, tenant)
+	if err != nil {
+		return nil, err
+	}
```
Comment (dragobt) [inline components/director/internal/systemfetcher/service.go:221]: No unit test for this case.

File: tests/system-fetcher/tests/handler_test.go [hunk 9]
```diff
@@ -658,6 +744,8 @@ func TestSystemFetcherDuplicateSystemsForTwoTenants(t *testing.T) {
 	require.NotEmpty(t, template2.ID)
 
 	triggerSync(t, tenant.TestTenants.GetDefaultTenantID())
+	waitForApplicationsToBeProcessed(ctx, t, tenant.TestTenants.GetDefaultTenantID(), 2)
+	triggerSync(t, tenant.TestTenants.GetSystemFetcherTenantID())
```
Comment (dragobt) [inline tests/system-fetcher/tests/handler_test.go:748]: Do we need to wait for this sync as well?

**Review by dragobt — CHANGES_REQUESTED** [review]

File: tests/system-fetcher/tests/handler_test.go (hunk 9 above)
Comment (dzahariev) [reply tests/system-fetcher/tests/handler_test.go:748]: Yes, at it may proceed longer.

**Review by dragobt — APPROVED** [review]

**Pull request merged**